### PR TITLE
Make S3 website private; add S3 origin

### PR DESCRIPTION
**Pull request by jmcgeheeiv**
## What

In this PR I added these enhancements:

- Make the S3 website accessible only via CloudFront. You may prefer that I add a variable to make this optional. Let me know.
- Add S3 origins, which are different from the currently-supported custom origins
- Change `http_port` and `https_port` to numbers merely to dissuade observers from attempting to use `null`. `null` causes a very, very hard-to-diagnose error that does not reveal the line number.

I will further note that Terraform insists that the complete data structure appearing in `variables.tf` be specified for `custom_origin` and `s3_origin` variables. Thus `lookup()` with default values is immaterial.

## Why

The reasons are included in the list of enhancements above. I hope it is more readable that way.

### CODE REVIEW

**Comment by jmcgeheeiv** [comment]
As I reported in #107, when using an S3 website origin, redirection from `subdir/` to `subdir/index.html` works only when `aliases` is specified. I added a notation to that effect in the `website_enabled` documentation.

I am tempted to go further and make `aliases` a required variable, but that breaks backward compatibility.

**Comment by nitrocode** [comment]
@jmcgeheeiv can you fix the merge conflicts ?

Also, it looks like PR https://github.com/cloudposse/terraform-aws-cloudfront-s3-cdn/pull/108 put in the following

https://github.com/cloudposse/terraform-aws-cloudfront-s3-cdn/blob/0818b34a4926ee9cb722fd581c9b940c7f5c99c9/main.tf#L249-L255

How would that work with this PR ? does that PR make this one no longer relevant ?

File: main.tf
Comment (nitrocode) [inline main.tf:274]: what if both `var.custom_origins` and `var.s3_origins` are provided ? can a cloudfront distribution have both ? or are they mutually exclusive ?

File: main.tf
Comment (nitrocode) [inline main.tf:251]: How is the `custom_header` above different from the `var.custom_origins` `origin` block `custom_header` ?

How come the same `custom_header` isn't applied to the `var.s3_origins` `origin` block below ?

Comment (nitrocode) [inline main.tf:251]: How come the same custom_header isn't applied to the var.s3_origins origin block below ?

File: main.tf
Comment (jmcgeheeiv) [reply main.tf:297]: An S3 origin is for S3 buckets, and custom origins are for S3 websites or apps. Both can be used at the same time. I use both.

File: main.tf
Comment (jmcgeheeiv) [reply main.tf:277]: Good eye there, @nitrocode. The two `custom_header` sections must be merged. I have a solution written, but I must first upgrade Terraform in order to test.

Comment (jmcgeheeiv) [reply main.tf]: Tested the merged `custom_header` sections with Terraform 0.13 and 0.14.

**Review by nitrocode — APPROVED** [review]

**Comment by jmcgeheeiv** [comment]
Okay @nitrocode, the code is tested with Terraform 0.13 and 0.14. The changes are in a single commit. I hope you can get this through before too many more commits come through.

**Comment by nitrocode** [comment]
It's not up to me. I'm not one of the approvers or in the engineering group. You'll have to be vocal in #pr-reviews channel to get this pr merged

**Comment by nitrocode** [comment]
Please fix the tests. You need to add the random provider with a version pin

**Comment by nitrocode** [comment]
There are also a lot of prs going in and a recent bug i introduced. Let's get all the other stuff merged and come back to this one and then fix the conflicts. Otherwise you'll be driving yourself up a wall like me after every merge

**Comment by jmcgeheeiv** [comment]
Okay, holding until I hear from you, @nitrocode.

**Comment by nitrocode** [comment]
The bug was fixed in a now closed PR. Feel free to correct tests and fix conflicts. After you do, I'd post in #pr-reviews in slack to get attention to it.

**Comment by Gowiem** [comment]
@jmcgeheeiv looks like this one is close, but two things outstanding: 

1. Bats tests are failing on the provider not being pinned properly: 

```
not ok 6 check if terraform providers are properly pinned
# (in test file test/.test-harness/test/terraform/provider-pinning.bats, line 23)
#   `terraform-config-inspect --json . | jq '.required_providers | to_entries[] | "  - \(.key)|\(.value.version_constraints[])"' > $TMPFILE' failed with status 5
# /usr/local/bin/terraform-config-inspect
# jq: error (at <stdin>:1197): Cannot iterate over null (null)
not ok 7 check if terraform providers have explicit source locations for TF =>0.13
# (in test file test/.test-harness/test/terraform/provider-pinning.bats, line 55)
#   `exit 1' failed
# /usr/local/bin/terraform-config-inspect
# 
# * Cloud Posse requires all providers to use registry format introduced in Terraform 0.13, for example
#     aws = {
#        source  = "hashicorp/aws"
#        version = ">= 3.0"
#     }
# 
# * Please add constraints for these providers:
#   - random
```

2. You've got merge conflicts in the README. They seem easily so should be quick to knock out. 

Get those updated and we'll give this a final pass and move it forward 🤙

**Comment by Gowiem** [comment]
@jmcgeheeiv something got gunked up with your PR / commits, so now we're seeing a changeset from a number of recent PRs instead of just your changes.

Sorry for this but can you do something along the following lines to get things back in order?

1. Check out a fresh `origin/master`
2. Create new branch
3. `cherry-pick` your commits onto that new branch
4. rename your old branch 
5. name your new branch like your branch
6. force-push your new branch to your forked `master`

Feel free to go about that differently if your git-fu is better than mine (not hard obviously).

**Comment by jmcgeheeiv** [comment]
I force-pushed a better version.

@Gowiem I think I see the error I made. Last time, I did something similar to the six step procedure you describe. However, the local working `master` branch remains out of sync with `origin/master`. Therefore your procedure and I both need to remember to also update the local `master` in preparation for next time:

```shell
git checkout master
git pull --force origin master
```

**Review by Gowiem — APPROVED** [review]
LGTM 👍

**Comment by Gowiem** [comment]
@jmcgeheeiv these bats tests are still failing -- Mind looking into them?

```
not ok 6 check if terraform providers are properly pinned
# (in test file test/.test-harness/test/terraform/provider-pinning.bats, line 23)
#   `terraform-config-inspect --json . | jq '.required_providers | to_entries[] | "  - \(.key)|\(.value.version_constraints[])"' > $TMPFILE' failed with status 5
# /usr/local/bin/terraform-config-inspect
# jq: error (at <stdin>:1175): Cannot iterate over null (null)
not ok 7 check if terraform providers have explicit source locations for TF =>0.13
# (in test file test/.test-harness/test/terraform/provider-pinning.bats, line 55)
#   `exit 1' failed
# /usr/local/bin/terraform-config-inspect
# 
# * Cloud Posse requires all providers to use registry format introduced in Terraform 0.13, for example
#     aws = {
#        source  = "hashicorp/aws"
#        version = ">= 3.0"
#     }
# 
# * Please add constraints for these providers:
#   - random
# 
# 
```

**Comment by jmcgeheeiv** [comment]
@Gowiem, am I correct in assuming that chatops `/test all` works only for CloudPosse personnel?

**Comment by Gowiem** [comment]
@jmcgeheeiv that's correct 😄  Leads to some confusion, but I believe it's done because Cloud Posse is spending money on an AWS bill out of its own pocket each time tests are run so they like to guard the fact that it could potentially be abused and they'd end up paying the full bill.

**Review by Gowiem — APPROVED** [review]
I believe this looks good, but I want to get a second set of eyes on it as well before we call it merged since it's a bigger PR. @nitrocode you mind giving a second approval?

**Comment by jmcgeheeiv** [comment]
Quick, before cloudpossebot changes its mind!

**Comment by jmcgeheeiv** [comment]
@Gowiem, yes, I was thinking that since `/test all` is not free, not everyone should be able to do it. Quite reasonable.

**Comment by jmcgeheeiv** [comment]
@3h4x @woz5999 please review while it's clean and green!

**Comment by woz5999** [comment]
the code changes lgtm but i've not super familiar with the implications of the changes, so i'd prefer to punt the actual merge to someone else.

**Comment by Gowiem** [comment]
@woz5999 haha I'm in the same boat and wanted another contributor to help me share the merge burden due to the large set of changes, but I shouldn't try to put that on others. I'll merge and deal with the fallout if this blows up this popular module. @jmcgeheeiv I'll be pinging you if that happens 😉

**Comment by Gowiem** [comment]
Thanks for the contribution @jmcgeheeiv (and working through our fun hoops on this one in particular)! Released as [0.61.0](https://github.com/cloudposse/terraform-aws-cloudfront-s3-cdn/releases/tag/0.61.0)!

**Pull request merged**
